Thumbnail integration tests

rgozim opened the pull request:
# What this PR does

Adds import options to AbstractServerTest base class and integration tests for Thumbnail loading.

rgozim committed:
Moved tests

rgozim committed:
Added a couple of tests

rgozim committed:
Added tests ThumbnailBeanStoreTest
Added ImportConfig to AbstractServerTest.importFile
Removed unused boolean from importFile in AbstractServerTest

rgozim committed:
Added tests ThumbnailBeanStoreTest

rgozim committed:
Added steps to javadocs

joshmoore commented:
Something about this branch looks _very_ weird.

rgozim commented:
Yea, that was strange. Should be better now.

rgozim committed:
Updated multi user tests to account for different group permissions

rgozim committed:
Small changes to rename methods and inner class of thumbnail tests

rgozim commented:
> Okay to move it if need be but maybe preserve the blank line before the OME imports?

I'll add it back again in the next commit. I'll be splitting the file to two test files soon anyway, so imports will probably change again.

rgozim committed:
Split thumbnail tests up into multiple files
Added conditions for isAdmin, isGroupOwner in PermissionsTest

rgozim committed:
Added test for large file import (to trigger pyramid generation)

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/SkipThumbnailsPermissionsTest.java:
```diff
@@ -260,22 +308,91 @@ public void testGetThumbnailWithRenderingSettingsChange(String permissions, bool
-        String format = ModelMockFactory.FORMATS[0];
-        File file = createImageFile(format);
-        return importFile(config, file, format).get(0);
+        BufferedImage bi = new BufferedImage(4096, 4096, BufferedImage.TYPE_INT_RGB);
```
Perhaps an empty `big-image&sizeX=4096&sizeY=4096.fake` file would suffice?

jburel replied:
it will

rgozim committed:
Added method for triggering pyramid generation
Passing tests for tests involving both owner and viewer of image with
thumbnails

jburel commented on components/tools/OmeroJava/test/integration/thumbnail/BatchLoadingTest.java:
```diff
@@ -0,0 +1,148 @@
+            tnCount++;
+        }
+        Assert.assertEquals(thumbNailCount, tnCount);
+        svc.close();
```
service will not be closed at that stage if the test failed
the test should be split into 2, one for each method invoked

jburel commented on components/tools/OmeroJava/test/integration/thumbnail/SingleFileTest.java:
```diff
@@ -0,0 +1,160 @@
+        byte[] lsValues = svc.getThumbnailByLongestSide(omero.rtypes
+                .rint(sizeX));
+        Assert.assertNotNull(lsValues);
+        Assert.assertTrue(lsValues.length > 0);
```
check that the thumbnail is of the expected size

jburel reviewed (commented):
We need to check that the correct thumbnailstore is used, since we have few store initialized in methods and they are not closed.

jburel committed:
Update doc

jburel committed:
Merge remote-tracking branch 'origin/develop' into thumbnail_integration_tests

jburel committed:
Add header

jburel committed:
Review error handling and close service

jburel committed:
Fix test

jburel committed:
Use fake Image for pyramid generation

jburel committed:
Use fake file and delete them on exit

jburel committed:
Check the size of image

jburel committed:
Delete tmp file

jburel committed:
Add method to check size in Utils

jburel committed:
Use common method

jburel committed:
Check image size

rgozim commented:
Lists of tests failing: 

![screenshot 2018-11-08 at 11 22 03](https://user-images.githubusercontent.com/3717090/48195842-9ca4da00-e348-11e8-9f1d-d6ace0bb3dc1.png)

`testGetThumbnailWithRenderingSettingsChange` fails with permission other than _rw----_.

Stacktrace is the same for each failed test, line 388 `re.resetDefaultSettings`:

```
omero.MissingPyramidException
    serverStackTrace = "ome.conditions.MissingPyramidException: Missing pyramid:/Users/rgozim/Projects/omero/data/Pixels/226_pyramid
                        	at ome.io.nio.SimpleBackOff.throwMissingPyramidException(SimpleBackOff.java:94)
                        	at ome.io.nio.PixelsService.handleMissingPyramid(PixelsService.java:778)
                        	at ome.io.nio.PixelsService._getPixelBuffer(PixelsService.java:583)
                        	at ome.io.nio.PixelsService.getPixelBuffer(PixelsService.java:541)
                        	at ome.logic.RenderingSettingsImpl.resetDefaults(RenderingSettingsImpl.java:519)
                        	at ome.logic.RenderingSettingsImpl.resetDefaults(RenderingSettingsImpl.java:1429)
                        	at ome.services.RenderingBean$11.doWork(RenderingBean.java:2196)
                        	at sun.reflect.GeneratedMethodAccessor301.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:573)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy72.doWork(Unknown Source)
                        	at ome.services.util.Executor$Impl.execute(Executor.java:452)
                        	at ome.services.util.Executor$Impl.execute(Executor.java:396)
                        	at ome.services.RenderingBean._resetDefaults(RenderingBean.java:2193)
                        	at ome.services.RenderingBean.internalReset(RenderingBean.java:713)
                        	at ome.services.RenderingBean.resetDefaultSettings(RenderingBean.java:677)
                        	at sun.reflect.GeneratedMethodAccessor1222.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy110.resetDefaultSettings(Unknown Source)
                        	at sun.reflect.GeneratedMethodAccessor1222.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:93)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy110.resetDefaultSettings(Unknown Source)
                        	at sun.reflect.GeneratedMethodAccessor1236.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:172)
                        	at ome.services.throttling.Callback.run(Callback.java:56)
                        	at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
                        	at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:140)
                        	at ome.services.blitz.impl.RenderingEngineI.resetDefaultSettings_async(RenderingEngineI.java:380)
                        	at sun.reflect.GeneratedMethodAccessor1235.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at omero.cmd.CallContext.invoke(CallContext.java:85)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy112.resetDefaultSettings_async(Unknown Source)
                        	at omero.api._RenderingEngineTie.resetDefaultSettings_async(_RenderingEngineTie.java:316)
                        	at omero.api._RenderingEngineDisp.___resetDefaultSettings(_RenderingEngineDisp.java:2017)
                        	at omero.api._RenderingEngineDisp.__dispatch(_RenderingEngineDisp.java:2437)
                        	at IceInternal.Incoming.invoke(Incoming.java:221)
                        	at Ice.ConnectionI.invokeAll(ConnectionI.java:2536)
                        	at Ice.ConnectionI.dispatch(ConnectionI.java:1145)
                        	at Ice.ConnectionI.message(ConnectionI.java:1056)
                        	at IceInternal.ThreadPool.run(ThreadPool.java:395)
                        	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
                        	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:832)
                        	at java.lang.Thread.run(Thread.java:748)
                        "
    serverExceptionClass = "ome.conditions.MissingPyramidException"
    message = "Missing pyramid:/Users/rgozim/Projects/omero/data/Pixels/226_pyramid"
    backOff = 1000
    pixelsID = 226
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at IceInternal.BasicStream.createUserException(BasicStream.java:2779)
	at IceInternal.BasicStream.access$300(BasicStream.java:14)
	at IceInternal.BasicStream$EncapsDecoder10.throwException(BasicStream.java:3298)
	at IceInternal.BasicStream.throwException(BasicStream.java:2291)
	at IceInternal.OutgoingAsync.throwUserException(OutgoingAsync.java:399)
	at omero.api.RenderingEnginePrxHelper.end_resetDefaultSettings(RenderingEnginePrxHelper.java:7916)
	at omero.api.RenderingEnginePrxHelper.resetDefaultSettings(RenderingEnginePrxHelper.java:7795)
	at omero.api.RenderingEnginePrxHelper.resetDefaultSettings(RenderingEnginePrxHelper.java:7782)
	at integration.thumbnail.SkipThumbnailsPermissionsTest.testGetThumbnailWithRenderingSettingsChange(SkipThumbnailsPermissionsTest.java:388)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:80)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
	at org.testng.SuiteRunner.run(SuiteRunner.java:240)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1198)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1123)
	at org.testng.TestNG.run(TestNG.java:1031)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)
```

jburel committed:
review failing tests

jburel committed:
Clean up

jburel committed:
Add extra test

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/BatchLoadingTest.java:
```diff
@@ -0,0 +1,172 @@
+        svc.close();
+    }
+
+
```
single blank line suffices

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/BatchLoadingTest.java:
```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2018 University of Dundee & Open Microscopy Environment.
```
There's certainly older code here.

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/SingleFileTest.java:
```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2018 University of Dundee & Open Microscopy Environment.
```
There's certainly older code here.

mtbc commented:
For each test method add Javadoc saying what it tests.

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/SkipThumbnailsPermissionsTest.java:
```diff
@@ -0,0 +1,640 @@
+        } catch (omero.ResourceError | omero.ReadOnlyGroupSecurityViolation e) {
+            // With permission rw----, the image is private to user 1.
+            // Expect this to fail for user 2.
+            if (!permissions.equalsIgnoreCase("rw----")) {
```
Why are all of these ignoring case?

jburel replied:
b/c in that case the operation is not permitted and the error is expected. This could be replaced by success message

mtbc replied:
I'm completely lost, sorry. The error would make a simple `equals` insufficient?

jburel replied:
I have reviewed the tests 
hopefully it is improved

mtbc commented:
Javadoc for util methods would be good too, `checkSize()` is a bit vague.

jburel committed:
Adjust date since files contain older code

jburel committed:
Remove extra line

jburel committed:
Add javadoc and use correct variable

jburel committed:
Add javadoc

jburel committed:
Disconnect

jburel committed:
Add javadoc

jburel committed:
Review test

jburel committed:
Review permissions and error handling

jburel committed:
Review doc

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/SkipThumbnailsPermissionsTest.java:
```diff
@@ -0,0 +1,686 @@
+        ThumbnailStorePrx svc = factory.createThumbnailStore();
+        Utils.setThumbnailStoreToPixels(svc, pixels.getId().getValue());
+        Utils.getThumbnail(svc);
+        svc.close();
```
Should this be in a `finally` as in the previous method? Also more below may need tweaking.

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/SkipThumbnailsPermissionsTest.java:
```diff
@@ -0,0 +1,686 @@
+        svc.close();
+
+        // Check that the thumbnails are different
+        Assert.assertFalse(Arrays.equals(user1Thumbnail, user2Thumbnail));
```
Should be an array-aware `assertEquals`?

jburel replied:
What we need will be an ``AssertNotEquals`` but the method takes Object so I don't think it will give what we need

mtbc replied:
Indeed you're correct: interesting! Never noticed there was more `assertEquals` than `assertNotEquals`.

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/SkipThumbnailsPermissionsTest.java:
```diff
@@ -0,0 +1,686 @@
+
+        // If user is an admin, add them to the system group
+        if (isAdmin) {
+            ExperimenterGroup systemGroup = new ExperimenterGroupI(iAdmin.getSecurityRoles().systemGroupId, false);
```
superclass has `roles` field

mtbc commented on components/tools/OmeroJava/test/integration/thumbnail/SkipThumbnailsPermissionsTest.java:
```diff
@@ -0,0 +1,686 @@
+        return f;
+    }
+
+    private void triggerPyramidGeneration(long pixelsId) throws ServerError {
```
@joshmoore: I suggested this approach when you were on leave. Absolutely open to less hacky ideas.

joshmoore replied:
Unless we extend the methods in `omero/cmd/FS.ice` this is probably find for a test (and is equivalent to loading the image itself)

jburel replied:
Thanks

mtbc commented:
[OMERO-DEV-merge-integration-java](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-integration-java/) has been green for a little while.

jburel committed:
Add try/finally

jburel committed:
Use roles field

jburel commented:
Let's check if the tests remain green tomorrow then we should be good for this first round

rgozim committed:
Updated some javadoc formatting

jburel commented:
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-integration-java/ is green 

Any objections to merge that first round of tests
